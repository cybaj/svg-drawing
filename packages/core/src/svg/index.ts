export * from './svg'
export * from './command'
export * from './point'
export * from './path'
