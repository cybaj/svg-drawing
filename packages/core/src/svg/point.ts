import type { PointObject } from '../types'

/**
 * Maximum number of decimal places is 5
 *
 * @todo Changeable parameter
 */
const coeffecient = 100000
const roundNumber = (num: number) => Math.round(num * coeffecient) / coeffecient

export class Point implements PointObject {
  private _x: number
  private _y: number
  constructor(x: number, y: number) {
    this._x = roundNumber(x)
    this._y = roundNumber(y)
  }

  public get x(): number {
    return this._x
  }

  public get y(): number {
    return this._y
  }

  public get degrees(): number {
    return Math.atan2(this._y, this._x) * (180 / Math.PI)
  }

  public get absoluteValue(): number {
    return Math.sqrt(Math.pow(this._x, 2.0) + Math.pow(this._y, 2.0))
  }

  public scale(r: number): Point {
    return new Point(this._x * r, this._y * r)
  }

  public scaleX(r: number): Point {
    return new Point(this._x * r, this._y)
  }

  public scaleY(r: number): Point {
    return new Point(this._x, this._y * r)
  }

  public add(p: PointObject): Point {
    return new Point(this._x + p.x, this._y + p.y)
  }

  public sub(p: PointObject): Point {
    return new Point(this._x - p.x, this._y - p.y)
  }

  public translate(p: PointObject): void {
    this._x += p.x
    this._y += p.y
  }

  public eql(p: PointObject): boolean {
    return this._x === p.x && this._y === p.y
  }

  public clone(): Point {
    return new Point(this._x, this._y)
  }

  public toJson(): PointObject {
    return {
      x: this._x,
      y: this._y,
    }
  }

  public static fromVector({
    value,
    degrees,
  }: {
    value: number
    degrees: number
  }) {
    const theta = (degrees * Math.PI) / 180
    return new Point(Math.cos(theta) * value, Math.sin(theta) * value)
  }
}
