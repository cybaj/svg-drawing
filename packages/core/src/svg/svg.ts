import type { SvgObject, SvgOption } from '../types'
import type { Path } from './path'

export class Svg {
  public paths: Path[]
  public width: number
  public height: number
  public background?: string

  constructor({ width, height, background }: SvgOption) {
    this.paths = []
    this.width = width
    this.height = height
    this.background = background
  }

  /** @todo Check height */
  public resize({ width, height }: { width: number; height: number }) {
    this.scale(width / this.width)
    this.width = width
    this.height = height
  }

  public scale(r: number): this {
    if (r !== 1) {
      for (let i = 0; i < this.paths.length; i += 1) {
        this.paths[i].scale(r)
      }
    }
    return this
  }

  public scaleX(r: number): this {
    if (r !== 1) {
      for (let i = 0; i < this.paths.length; i += 1) {
        this.paths[i].scaleX(r)
      }
    }
    return this
  }

  public scaleY(r: number): this {
    if (r !== 1) {
      for (let i = 0; i < this.paths.length; i += 1) {
        this.paths[i].scaleY(r)
      }
    }
    return this
  }

  public addPath(pa: Path | Path[]): this {
    if (Array.isArray(pa)) {
      this.paths.push(...pa)
    } else {
      this.paths.push(pa)
    }
    return this
  }

  public deletePath(deletePath: Path): this {
    this.paths = this.paths.filter((path) => path.key !== deletePath.key)
    return this
  }

  public clonePaths(): Path[] {
    return this.paths.map((p) => p.clone())
  }

  public toJson(): SvgObject {
    return {
      width: this.width,
      height: this.height,
      background: this.background,
      paths: this.paths.map((p) => p.toJson()),
    }
  }

  public copy(svg: any extends Svg ? Svg : never): this {
    this.paths = svg.clonePaths()
    if (svg.width && this.width) {
      this.scale(this.width / svg.width)
    }
    return this
  }

  public clone(): Svg {
    const svg = new Svg({
      width: this.width,
      height: this.height,
      background: this.background,
    })
    svg.addPath(this.clonePaths())
    return svg
  }
}
