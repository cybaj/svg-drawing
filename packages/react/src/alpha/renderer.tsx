import React, { useCallback, useMemo, useState, useEffect } from 'react'
import {
  Path,
  PathObject,
  PointObject,
  ControlPoint,
  BoundingBox,
  EditPath as EditPathCore,
} from '@svg-drawing/core'
import {
  EditPathProps,
  EditSvgProps,
  SelectPathHandler,
  SvgProps,
  EditPathIndex,
} from './types'

export const Svg = ({
  background,
  paths,
  width,
  height,
  ...rest
}: SvgProps) => (
  <svg width={width} height={height} {...rest}>
    {background && <rect width={width} height={height} fill={background} />}
    {paths.map((pathAttr, i) => (
      <path key={i} {...pathAttr} />
    ))}
  </svg>
)

export const EditSvg = ({
  editing,
  onSelect: handleSelect,
  onUpdate: handleUpdate,
  onCancel: handleCancel,
  background,
  paths,
  width,
  height,
  ...rest
}: EditSvgProps) => {
  const handleSelectPath = useCallback(
    (pathIndex: number): SelectPathHandler => (arg) => {
      handleSelect({
        path: pathIndex,
        ...arg,
      })
    },
    [handleSelect]
  )

  return (
    <svg width={width} height={height} {...rest}>
      {background && <rect width={width} height={height} fill={background} />}
      {paths.map((pathAttr: PathObject, i) => (
        <EditPath
          key={i}
          {...(pathAttr as any)}
          editingPath={
            i === editing?.path
              ? {
                  command: editing?.command,
                  value: editing?.value,
                }
              : null
          }
          onUpdate={handleUpdate}
          onCancel={handleCancel}
          onSelectPath={handleSelectPath(i)}
        />
      ))}
    </svg>
  )
}

const EDIT_CONFIG = {
  line: 1,
  point: 3,
  color: {
    main: '#09f',
    sub: '#f90',
    selected: '#f00',
  },
  fill: {
    default: 'none',
    boundingBox: 'rgba(0,0,0,0)',
    selected: 'rgba(0,0,0,0.1)',
  },
} as const

const EditPath = ({
  d,
  editingPath,
  onSelectPath: handleSelectPath,
  onUpdate: handleUpdate,
  onCancel: handleCancel,
  ...attrs
}: EditPathProps & PathObject) => {
  const [currentPosition, setCurrentPosition] = useState<PointObject | null>(
    null
  )
  const [moveBoundingBox, setMoveBoundingBox] = useState(false)

  const editPath: EditPathCore = useMemo(() => {
    const p = new Path()
    if (d) p.parseCommandString(d)
    return new EditPathCore(p)
  }, [d])

  const controlPoints: ControlPoint[] = useMemo(() => editPath.controlPoints, [
    editPath,
  ])

  const boundingBox: BoundingBox = useMemo(() => editPath.boundingBox, [
    editPath,
  ])

  const handleClickPath = useCallback(
    (ev: React.MouseEvent<HTMLOrSVGElement>) => {
      setCurrentPosition(null)
      handleSelectPath({})
    },
    [handleSelectPath]
  )

  const handleMouseMoveCircle = useCallback(
    (ev: React.MouseEvent<SVGCircleElement>) => {
      if (!currentPosition) return
      handleUpdate({
        x: ev.clientX - currentPosition.x,
        y: ev.clientY - currentPosition.y,
      })
      setCurrentPosition({
        x: ev.clientX,
        y: ev.clientY,
      })
    },
    [currentPosition, handleUpdate]
  )

  const handleSelectedCircle = useCallback(
    (selector: EditPathIndex) => (ev: React.MouseEvent<SVGCircleElement>) => {
      handleSelectPath(selector)
      setCurrentPosition({
        x: ev.clientX,
        y: ev.clientY,
      })
    },
    [handleSelectPath]
  )

  const isSelectedPoint = useCallback(
    (index: EditPathIndex): boolean => {
      if (!editingPath) return false
      return (
        editingPath.command === index.command &&
        editingPath.value === index.value
      )
    },
    [editingPath]
  )

  const isSelectedBoundingBox = useMemo(() => {
    if (!editingPath) return false
    return editingPath.command === undefined
  }, [editingPath])

  const handleMoveStartBoundingBox = useCallback(
    (ev: React.MouseEvent<SVGRectElement>) => {
      handleSelectPath({})
      setCurrentPosition({
        x: ev.clientX,
        y: ev.clientY,
      })
      setMoveBoundingBox(true)
    },
    [handleSelectPath]
  )
  const handleMoveEndBoundingBox = useCallback(
    (ev: MouseEvent) => {
      if (!moveBoundingBox || !isSelectedBoundingBox || !currentPosition) return
      setMoveBoundingBox(false)
      handleUpdate({
        x: ev.clientX - currentPosition.x,
        y: ev.clientY - currentPosition.y,
      })
    },
    [moveBoundingBox, isSelectedBoundingBox, currentPosition, handleUpdate]
  )

  useEffect(() => {
    window.addEventListener('mouseup', handleMoveEndBoundingBox)
    return () => window.removeEventListener('mouseup', handleMoveEndBoundingBox)
  }, [handleMoveEndBoundingBox])
  if (!editingPath) return <path d={d} {...attrs} onClick={handleClickPath} />
  return (
    <>
      <path d={d} {...attrs} />
      <path
        d={d}
        onClick={handleClickPath}
        strokeWidth={EDIT_CONFIG.line}
        stroke={EDIT_CONFIG.color.main}
        fill={EDIT_CONFIG.fill.default}
      />
      <rect
        {...boundingBox}
        stroke={EDIT_CONFIG.color.main}
        fill={
          isSelectedBoundingBox
            ? EDIT_CONFIG.fill.selected
            : EDIT_CONFIG.fill.boundingBox
        }
        onMouseDown={handleMoveStartBoundingBox}
        // onMouseMove={handleMoveBoundingBox}
      />
      {controlPoints.map(({ points, d }: ControlPoint, commandIndex) => (
        <g key={commandIndex}>
          <path
            d={d}
            strokeWidth={EDIT_CONFIG.line}
            stroke={EDIT_CONFIG.color.main}
            fill={EDIT_CONFIG.fill.default}
          />
          {points.map((po: PointObject, k) => {
            const valueIndex = k * 2
            const editPathIndex = {
              command: commandIndex,
              value: valueIndex,
            }
            return (
              <circle
                key={k}
                cx={po.x}
                cy={po.y}
                onMouseDown={handleSelectedCircle(editPathIndex)}
                onMouseMove={
                  isSelectedPoint(editPathIndex)
                    ? handleMouseMoveCircle
                    : undefined
                }
                r={EDIT_CONFIG.point}
                style={{
                  fill: isSelectedPoint(editPathIndex)
                    ? EDIT_CONFIG.color.selected
                    : EDIT_CONFIG.color.sub,
                }}
              />
            )
          })}
        </g>
      ))}
    </>
  )
}
