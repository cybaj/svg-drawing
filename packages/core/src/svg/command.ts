import { CommandType, PointObject } from '../types'
import { roundUp } from '../utils'
import { Point } from './point'

export const COMMAND_TYPE: { [name: string]: CommandType } = {
  MOVE: 'M', // M 0 0
  MOVE_RELATIVE: 'm', // m 0 0
  LINE: 'L', // L 1 1
  LINE_RELATIVE: 'l', // l 1 1
  CUBIC_BEZIER_CURVE: 'C', // C 1 1 2 2 3 3
  CUBIC_BEZIER_CURVE_RELATIVE: 'c', // c 1 1 2 2 3 3
  CLOSE: 'Z', // Z, z
  HORIZONTAL: 'H', // H 10
  HORIZONTAL_RELATIVE: 'h', // h 10
  VERTICAL: 'V', // V 20
  VERTICAL_RELATIVE: 'v', // v 20
  ARC_CURVE: 'A', // A 6 4 10 0 1 14 10
  ARC_CURVE_RELATIVE: 'a', // A 6 4 10 0 1 14 10
  QUADRATIC_CURVE: 'Q', // Q 10 60 10 30
  QUADRATIC_CURVE_RELATIVE: 'q', // q 10 60 10 30
  SHORTCUT_CURVE: 'S', // S 10 60 10 30
  SHORTCUT_CURVE_RELATIVE: 's', // s 10 60 10 30
} as const

// TODO: compatible COMMAND_TYPE
export class Command {
  public type: CommandType
  public value: number[]
  // TODO: Convert data format to number array.
  constructor(type: CommandType, value: number[] = []) {
    this.value = value
    this.type = type
  }

  public set cr(po: Point | undefined) {
    if (!po) return
    if (!this.isCubicBezierCurve) {
      return
    }
    this.value.splice(2, 1, po.x)
    this.value.splice(3, 1, po.y)
  }
  public get cr(): Point | undefined {
    if (!this.isCubicBezierCurve) {
      return undefined
    }
    const [x, y] = this.value.slice(2, 4)
    return new Point(x, y)
  }
  public set cl(po: Point | undefined) {
    if (!po) return
    if (!this.isCurve) {
      return
    }
    this.value.splice(0, 1, po.x)
    this.value.splice(1, 1, po.y)
  }
  public get cl(): Point | undefined {
    if (!this.isCurve) {
      return undefined
    }
    const [x, y] = this.value.slice(0, 2)
    return new Point(x, y)
  }

  public set point(po: Point | undefined) {
    if (!po) return
    this.value.splice(this.value.length - 2, 1, po.x)
    this.value.splice(this.value.length - 1, 1, po.y)
  }
  public get point(): Point | undefined {
    const xy = this.value.slice(this.value.length - 2)
    return xy.length === 2 ? new Point(xy[0], xy[1]) : undefined
  }

  public toString(): string {
    if (this.type === COMMAND_TYPE.CLOSE) return COMMAND_TYPE.CLOSE
    return `${this.type} ${this.value.map((v) => roundUp(v)).join(' ')}`
  }

  public scale(r: number): Command {
    const upd = new Command(
      this.type,
      this.value.map((p) => p * r)
    )
    return upd
  }

  public scaleX(r: number): Command {
    const point = this.point?.scaleX(r)
    const cl = this.cl?.scaleX(r)
    const cr = this.cr?.scaleX(r)
    return new Command(
      this.type,
      [cl, cr, point].reduce(
        (res: number[], po: Point | undefined) =>
          po ? [...res, po.x, po.y] : res,
        []
      )
    )
  }

  public scaleY(r: number): Command {
    const point = this.point?.scaleY(r)
    const cl = this.cl?.scaleY(r)
    const cr = this.cr?.scaleY(r)
    return new Command(
      this.type,
      [cl, cr, point].reduce(
        (res: number[], po: Point | undefined) =>
          po ? [...res, po.x, po.y] : res,
        []
      )
    )
  }

  public clone(): Command {
    return new Command(this.type, this.value.slice())
  }

  public get isCubicBezierCurve(): boolean {
    switch (this.type) {
      case COMMAND_TYPE.CUBIC_BEZIER_CURVE:
      case COMMAND_TYPE.CUBIC_BEZIER_CURVE_RELATIVE:
        return this.value.length === 6
      default:
        return false
    }
  }
  public get isCurve(): boolean {
    switch (this.type) {
      case COMMAND_TYPE.CUBIC_BEZIER_CURVE:
      case COMMAND_TYPE.CUBIC_BEZIER_CURVE_RELATIVE:
        return this.value.length === 6
      case COMMAND_TYPE.QUADRATIC_CURVE:
      case COMMAND_TYPE.QUADRATIC_CURVE_RELATIVE:
      case COMMAND_TYPE.SHORTCUT_CURVE:
      case COMMAND_TYPE.SHORTCUT_CURVE_RELATIVE:
        return this.value.length === 4
      default:
        return false
    }
  }

  public get isAbsolute(): boolean {
    return [
      COMMAND_TYPE.MOVE,
      COMMAND_TYPE.LINE,
      COMMAND_TYPE.CUBIC_BEZIER_CURVE,
      COMMAND_TYPE.ARC_CURVE,
      COMMAND_TYPE.QUADRATIC_CURVE,
      COMMAND_TYPE.SHORTCUT_CURVE_RELATIVE,
    ].includes(this.type)
  }
  public translate(p: PointObject) {
    if (!this.isAbsolute) return
    const po = new Point(p.x, p.y)
    this.point = this.point?.add(po)
    this.cr = this.cr?.add(po)
    this.cl = this.cl?.add(po)
  }
}
