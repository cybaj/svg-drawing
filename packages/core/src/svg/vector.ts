import { Point } from './point'

export class Vector {
  public angle: number
  public value: number
  constructor(v: number, a: number) {
    this.value = v
    this.angle = a
  }

  public toPoint(): Point {
    const x = Math.cos(this.angle) * this.value
    const y = Math.sin(this.angle) * this.value
    return new Point(x, y)
  }

  public scale(r: number): Vector {
    return new Vector(this.value * r, this.angle)
  }
}
