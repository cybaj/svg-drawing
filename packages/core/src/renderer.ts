import { Svg, SvgOption } from './svg'

export type RendererOption = Omit<SvgOption, 'width' | 'height'>
export class Renderer {
  public el: HTMLElement
  public top: number
  public left: number
  public svg: Svg
  constructor(el: HTMLElement, opt: RendererOption = {}) {
    const { width, height, left, top } = el.getBoundingClientRect()
    /**
     * Setup parameter
     */
    this.el = el
    this.left = left
    this.top = top
    this.svg = new Svg({ width, height, ...opt })
    el.appendChild(this.svg.toElement())
  }
  /**
   * render
   * TODO: XSS test
   */
  public update(): void {
    this.el.replaceChild(this.svg.toElement(), this.el.childNodes[0])
  }

  public resizeElement(
    param?:
      | DOMRect
      | { width: number; height: number; top: number; left: number }
  ): void {
    const { width, height, left, top } =
      param || this.el.getBoundingClientRect()
    // TODO: Resizing improve
    this.svg.scalePath(width / this.svg.width)
    this.svg.width = width
    this.svg.height = height
    this.left = left
    this.top = top
  }
}
