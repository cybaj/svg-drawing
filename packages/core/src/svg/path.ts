import { CommandType, PathObject, PointObject } from '../types'
import { kebab2camel } from '../utils'
import { Command, COMMAND_TYPE } from './command'

/**
 * TODO: refactor command.
 * The following commands are not supported. Cannot support commands that use `M` or` z` more than once
 * `M 0 0 L 1 1 Z M 1 1 L 2 2 Z`
 */
export class Path {
  public attrs: PathObject
  public commands: Command[]

  constructor({ d, ...attrs }: PathObject = {}) {
    this.attrs = attrs
    this.commands = []
    if (d) this.parseCommandString(d)
  }

  public scale(r: number): this {
    this.commands = this.commands.map((c: Command) => c.scale(r))
    this.attrs.strokeWidth = String(r * +(this.attrs.strokeWidth || 0))
    return this
  }

  public scaleX(r: number): this {
    this.commands = this.commands.map((c: Command) => c.scaleX(r))
    return this
  }

  public scaleY(r: number): this {
    this.commands = this.commands.map((c: Command) => c.scaleY(r))
    return this
  }

  public addCommand(param: Command | Command[]): this {
    if (Array.isArray(param)) {
      this.commands.push(...param)
    } else {
      this.commands.push(param)
    }
    return this
  }

  public deleteCommand(i: number): this {
    this.commands.splice(i, 1)
    return this
  }

  public getCommandString(): string {
    if (this.commands.length === 0) return ''
    return this.commands
      .map((com: Command, _i: number) => com.toString())
      .join(' ')
      .trim()
  }

  // TODO: Valid Command type
  public parseCommandString(d: string): void {
    this.commands = []
    let type: CommandType | null = null
    let value: number[] = []
    const c = d.split(' ')
    const checkType = (c: any): CommandType | null =>
      Object.values(COMMAND_TYPE).includes(c) ? c : null
    for (let i = 0; i < c.length; i += 1) {
      const t = checkType(c[i])
      // COMMAND Parse
      if (t) {
        if (!type) {
          type = t
          continue
        }
        this.commands.push(new Command(type, value))
        type = t
        value = []
        continue
      }
      if (isNaN(+c[i])) {
        return
      }
      value.push(+c[i])
    }

    if (type !== null) {
      this.commands.push(new Command(type, value))
    }
  }

  public parsePathElement(pEl: SVGPathElement): this {
    for (let i = 0; i < pEl.attributes.length; i += 1) {
      const attr: Attr | null = pEl.attributes.item(i)
      if (!attr || !attr.value) continue
      if (attr.name === 'd') {
        this.parseCommandString(attr.value)
        continue
      }
      this.attrs = {
        ...this.attrs,
        [kebab2camel(attr.name)]: attr.value,
      }
    }
    return this
  }

  public toJson(): PathObject {
    return {
      ...this.attrs,
      d: this.getCommandString(),
    }
  }

  public translate(po: PointObject): void {
    for (let i = 0; i < this.commands.length; i += 1) {
      this.commands[i].translate(po)
    }
  }

  public clone(): Path {
    const path = new Path(this.attrs)
    this.commands.map((c) => {
      path.commands.push(c.clone())
    })
    return path
  }
}
