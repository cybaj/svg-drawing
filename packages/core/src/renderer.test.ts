import { svgObjectToElement } from './renderer'
import { Svg, Path, Command } from './svg'

describe('renderer.ts', () => {
  const svg = new Svg({ width: 500, height: 500 })
    // TODO: rewrite bezier curve test
    .addPath(
      new Path()
        .addCommand(new Command('M', [0, 0]))
        .addCommand(new Command('C', [0.2, 0.2, 0.6, 0.8, 1, 1]))
        .addCommand(new Command('C', [1.4, 1.2, 1.6, 1.2, 2, 1]))
        .addCommand(new Command('C', [2.4, 0.8, 2.8, 0.2, 3, 0]))
    )
    .addPath(
      new Path({
        strokeLinecap: 'square',
        strokeLinejoin: 'mitter',
      })
        .addCommand(new Command('M', [4, 4]))
        .addCommand(new Command('L', [9, 4]))
        .addCommand(new Command('L', [9, 8]))
        .addCommand(new Command('L', [3, 0]))
        .addCommand(new Command('Z'))
    )

  it('svgObjectToElement', () => {
    const el = svgObjectToElement(svg.toJson())
    expect(el).toMatchSnapshot()
  })
})
