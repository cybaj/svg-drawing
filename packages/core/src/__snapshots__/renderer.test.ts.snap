// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renderer.ts svgObjectToElement 1`] = `
<svg
  height="500"
  version="1.1"
  width="500"
  xmlns="http://www.w3.org/2000/svg"
  xmlns:xlink="http://www.w3.org/1999/xlink"
>
  <path
    d="M0 0 C0.2 0.2 0.6 0.8 1 1 C1.4 1.2 1.6 1.2 2 1 C2.4 0.8 2.8 0.2 3 0"
  />
  <path
    d="M4 4 L9 4 L9 8 L3 0 z"
    stroke-linecap="square"
    stroke-linejoin="mitter"
  />
</svg>
`;
